### ISSUE TITLE
Workspaces - CVAT - Get Workflow Templates via SDK

### ISSUE THREAD

**Issue by rushtehrani** [onepanelio/core#387]
Currently, we're using environment variables, instead we should use the Python SDK's [list_workflow_templates](https://github.com/onepanelio/python-sdk/blob/master/docs/WorkflowTemplateServiceApi.md#list_workflow_templates) API to get a list of templates and only display the ones in CVAT with the label `used-by: cvat`

This will replace the need for `ONEPANEL_MASKRCNN_TEMPLATE_ID` and `ONEPANEL_OD_TEMPLATE_ID` environment variables.

**Comment by savan77** [comment]
@rushtehrani I think we'll have to dynamically generate a form for workflow parameters. Is there an API to get a list of parameters for a given workflow?

**Comment by rushtehrani** [comment]
@savan77 makes sense, there are two parts to this:

1. Get a list of workflow templates to display the available workflows for training. This is the SDK call listed above.
2. Once the user selects one of them, make another SDK call to get the parameters for that  particular WorkflowTemplate. I have to defer to @Vafilor on this one as I don’t recall which API call this is.

**Comment by savan77** [comment]
@rushtehrani To elaborate on the point that we discussed earlier, I was thinking about having a separate menu (just like Tasks, Models, and Analytics) for Onepanel. Maybe we can call it Workflows. There, we can have a `Execute workflow` button which will enlist all available workflows when user clicks on it. Once the user selects a particular workflow, we can get all available parameters for that workflow, assuming we have an API for the same. This is similar to what we have now, but once the workflow has been executed we can show that on that page just like a list of models on the Models page. We can also have a button on that list which links to the executed workflow. 

This will make it easier to keep track of executed workflows and its also generic and not related to model training only. Moreover, this way we will also have a separate component for Onepanel.

**Comment by aleksandrmelnikov** [comment]
Adding the following endpoint:
`/onepanelio/workflow_templates` inside of CVAT.
You will get the following response back:
```json
{
  "page": 1,
  "pages": 1,
  "total_count": 6,
  "count": 6,
  "workflow_templates": [
    {
      "is_archived": null,
      "name": "Test GPU",
      "versions": "11",
      "version": null,
      "created_at": "2020-07-03T01:13:35Z",
      "manifest": null,
      "cron_stats": null,
      "is_latest": null,
      "uid": "test-gpu",
      "modified_at": null,
      "stats": {
        "last_executed": "2020-07-17T20:18:10Z",
        "completed": 3,
        "total": 42,
        "running": 1,
        "terminated": 22,
        "failed": 15
      },
      "labels": null
    },
    {
      "is_archived": null,
      "name": "Tensorflow Sample Training",
      "versions": "2",
      "version": null,
      "created_at": "2020-05-27T01:35:00Z",
      "manifest": null,
      "cron_stats": null,
      "is_latest": null,
      "uid": "tensorflow-sample-training",
      "modified_at": null,
      "stats": {
        "last_executed": "2020-05-29T17:59:36Z",
        "completed": 2,
        "total": 3,
        "running": null,
        "terminated": null,
        "failed": 1
      },
      "labels": [
        {
          "key": "a",
          "value": "b"
        },
        {
          "key": "c",
          "value": "d"
        }
      ]
    },
    {
      "is_archived": null,
      "name": "PyTorch Sample Training",
      "versions": "3",
      "version": null,
      "created_at": "2020-05-27T00:02:29Z",
      "manifest": null,
      "cron_stats": null,
      "is_latest": null,
      "uid": "pytorch-sample-training",
      "modified_at": null,
      "stats": {
        "last_executed": "2020-06-11T05:06:03Z",
        "completed": 3,
        "total": 3,
        "running": null,
        "terminated": null,
        "failed": null
      },
      "labels": null
    },
    {
      "is_archived": null,
      "name": "TF Object Detection Training",
      "versions": "24",
      "version": null,
      "created_at": "2020-05-22T19:35:22Z",
      "manifest": null,
      "cron_stats": {
        "total": 1
      },
      "is_latest": null,
      "uid": "tf-object-detection-training",
      "modified_at": null,
      "stats": {
        "last_executed": "2020-07-20T15:58:28Z",
        "completed": 22,
        "total": 73,
        "running": 2,
        "terminated": 20,
        "failed": 29
      },
      "labels": null
    },
    {
      "is_archived": null,
      "name": "MaskRCNN Training",
      "versions": "28",
      "version": null,
      "created_at": "2020-05-15T22:53:28Z",
      "manifest": null,
      "cron_stats": null,
      "is_latest": null,
      "uid": "maskrcnn-training",
      "modified_at": null,
      "stats": {
        "last_executed": "2020-06-20T04:51:34Z",
        "completed": 2,
        "total": 42,
        "running": null,
        "terminated": 25,
        "failed": 15
      },
      "labels": null
    },
    {
      "is_archived": null,
      "name": "Quick DAG",
      "versions": "36",
      "version": null,
      "created_at": "2020-05-15T20:25:58Z",
      "manifest": null,
      "cron_stats": null,
      "is_latest": null,
      "uid": "quick-dag",
      "modified_at": null,
      "stats": {
        "last_executed": "2020-06-03T18:14:06Z",
        "completed": 38,
        "total": 63,
        "running": null,
        "terminated": 17,
        "failed": 8
      },
      "labels": null
    }
  ]
}
```

**Comment by aleksandrmelnikov** [comment]
This is ready to test.

You need onepanel-core to have a new image: `onepanel/core:01ff178`

If you want to test it as a workspace template, you can use this YAML.
```yaml
# Docker containers that are part of the Workspace
containers:
- name: cvat-db
  image: postgres:10-alpine
  env:
  - name: POSTGRES_USER
    value: root
  - name: POSTGRES_DB
    value: cvat
  - name: POSTGRES_HOST_AUTH_METHOD
    value: trust
  - name: PGDATA
    value: /var/lib/psql/data
  ports:
  - containerPort: 5432
    name: tcp
  volumeMounts:
  - name: db
    mountPath: /var/lib/psql
- name: cvat-redis
  image: redis:4.0-alpine
  ports:
  - containerPort: 6379
    name: tcp
- name: cvat
  image: onepanel/cvat:v0.7.11-core.389-6
  env:
  - name: DJANGO_MODWSGI_EXTRA_ARGS
    value: ""
  - name: ALLOWED_HOSTS
    value: '*'
  - name: CVAT_REDIS_HOST
    value: localhost
  - name: CVAT_POSTGRES_HOST
    value: localhost
  - name: CVAT_SHARE_URL
    value: /home/django/data
  - name: DJANGO_LOG_SERVER_HOST
    value: localhost
  - name: DJANGO_LOG_SERVER_PORT
    value: 5000
  - name: DJANGO_LOG_VIEWER_HOST
    value: localhost
  - name: DJANGO_LOG_VIEWER_PORT
    value: 5601
  - name: no_proxy
    value: cvat-kibana, cvat-logstash
  ports:
  - containerPort: 8080
    name: http
  volumeMounts:
  - name: data
    mountPath: /home/django/data
  - name: keys
    mountPath: /home/django/keys
  - name: logs
    mountPath: /home/django/logs
  - name: models
    mountPath: /home/django/models
  - name: share
    mountPath: /home/django/share
- name: cvat-ui
  image: onepanel/cvat-ui:v0.7.8
  ports:
  - containerPort: 80
    name: http
- name: filesyncer
  image: onepanel/filesyncer:v0.0.4
  command: ['python3', 'main.py']
  volumeMounts:
  - name: share
    mountPath: /mnt/share
- name: cvat-elasticsearch
  image: onepanel/cvat-elasticsearch:v0.0.1
  volumeMounts:
  - name: events
    mountPath: /usr/share/elasticsearch/data
  ports:
  - containerPort: 9200
    name: http
- name: cvat-kibana
  image: onepanel/cvat-kibana:v0.0.1
  ports:
  - containerPort: 5601
    name: http
  env:
  - name: ELASTICSEARCH_URL
    value: http://localhost:9200
- name: cvat-kibana-setup
  image:  onepanel/cvat:v0.7.11-core.389-6
  command: ['bash']
  args: ['-c','/bin/bash wait-for-it.sh localhost:9200 -t 0 --; /bin/bash wait-for-it.sh localhost:5601 -t 0 -- ; /usr/bin/python3 /tmp/components/analytics/kibana/setup.py -Hlocalhost /tmp/components/analytics/kibana/export.json ; sleep infinity']
    #no_proxy: cvat-elasticsearch, cvat-kibana, ${no_proxy}
  env:
  - name: DJANGO_LOG_SERVER_HOST
    value: localhost
  - name: DJANGO_LOG_SERVER_PORT
    value: 5000
  - name: DJANGO_LOG_VIEWER_HOST
    value: localhost
  - name: DJANGO_LOG_VIEWER_PORT
    value: 5601
- name: cvat-logstash
  image: onepanel/cvat-logstash:v0.0.2
  ports:
  - containerPort: 5000
    name: tcp
  

ports:
- name: cvat-ui
  port: 80
  protocol: TCP
  targetPort: 80
- name: cvat
  port: 8080
  protocol: TCP
  targetPort: 8080
- name: redis
  port: 6379
  protocol: TCP
  targetPort: 6379
- name: kibana
  port: 5601
  protocol: TCP
  targetPort: 5601
- name: elasticsearch
  port: 9200
  protocol: TCP
  targetPort: 9200

routes:
- match:
  - uri:
      regex: /api/.*|/onepanelio/.*|/git/.*|/tensorflow/.*|/auto_annotation/.*|/analytics/.*|/static/.*|/admin/.*|/documentation/.*|/dextr/.*|/reid/.*
  - queryParams:
      id:
        regex: \d+.*
  route:
  - destination:
      port:
        number: 8080
- match:
  - uri:
      prefix: /
  route:
  - destination:
      port:
        number: 80
# DAG Workflow to be executed once a Workspace action completes
# postExecutionWorkflow:
#   entrypoint: main
#   templates:
#   - name: main
#     dag:
#        tasks:
#        - name: slack-notify
#          template: slack-notify
#   - name: slack-notify
#     container:
#       image: technosophos/slack-notify
#       args:
#       - SLACK_USERNAME=onepanel SLACK_TITLE="Your workspace is ready" SLACK_ICON=https://www.gravatar.com/avatar/5c4478592fe00878f62f0027be59c1bd SLACK_MESSAGE="Your workspace is now running" ./slack-notify
#       command:
#       - sh
#       - -c
```

Longer version:
You will need to install a beta version of onepanel-sdk
`pip install onepanel-sdk==0.12.0b`

You will need to update the routes in the workspace YAML (if you don't copy the example above)
```yaml
routes:
- match:
  - uri:
      regex: /api/.*|/onepanelio/.*|/git/.*|/tensorflow/.*|/auto_annotation/.*|/analytics/.*|/static/.*|/admin/.*|/documentation/.*|/dextr/.*|/reid/.*
  - queryParams:
      id:
        regex: \d+.*
```

You will also need to use a CVAT image: `cvat:v0.7.11-core.389-6`

Lastly, the endpoints are:
```
/onepanelio/workflow_templates
/onepanelio/workflow_template/tensorflow-training/0
```

You can manually test these in the browser (easiest) or via CURL request.
```
https://authtoken3--default.alex.onepanel.io/onepanelio/workflow_template/tensorflow-training/0
https://authtoken3--default.alex.onepanel.io/onepanelio/workflow_templates
```

**Comment by aleksandrmelnikov** [comment]
Example JSON from `workflow_template/tensorflow-training/0`

```json
{
  "uid": "tensorflow-training",
  "cron_stats": null,
  "version": "1595562883661461978",
  "created_at": "2020-07-23T22:36:28Z",
  "modified_at": "2020-07-23T22:36:28Z",
  "is_latest": null,
  "name": "TensorFlow Training",
  "is_archived": null,
  "labels": null,
  "parameters": [
    {
      "type": null,
      "required": null,
      "name": "source",
      "options": null,
      "hint": null,
      "value": "name: command\norder: 1\nvalue: python mnist/main.py --epochs=5\n",
      "display_name": null
    },
    {
      "type": null,
      "required": null,
      "name": "command",
      "options": null,
      "hint": null,
      "value": "name: command\norder: 1\nvalue: python mnist/main.py --epochs=5\n",
      "display_name": null
    }
  ],
  "versions": "1",
  "stats": {
    "total": null,
    "running": null,
    "last_executed": "0001-01-01T00:00:00Z",
    "completed": null,
    "failed": null,
    "terminated": null
  },
  "manifest": "entrypoint: main\narguments:\n    parameters:\n    - name: source\n      value: https://github.com/onepanelio/tensorflow-examples.git\n    - name: command\n      value: \"python mnist/main.py --epochs=5\"\nvolumeClaimTemplates:\n  - metadata:\n      name: data\n    spec:\n      accessModes: [ \"ReadWriteOnce\" ]\n      resources:\n        requests:\n          storage: 2Gi\n  - metadata:\n      name: output\n    spec:\n      accessModes: [ \"ReadWriteOnce\" ]\n      resources:\n        requests:\n          storage: 2Gi\ntemplates:\n  - name: main\n    dag:\n      tasks:\n      - name: train-model\n        template: pytorch\n# Uncomment section below to send metrics to Slack\n#      - name: notify-in-slack\n#        dependencies: [train-model]\n#        template: slack-notify-success\n#        arguments:\n#          parameters:\n#          - name: status\n#            value: \"{{tasks.train-model.status}}\"\n#          artifacts:\n#          - name: metrics\n#            from: \"{{tasks.train-model.outputs.artifacts.sys-metrics}}\"\n  - name: pytorch\n    inputs:\n      artifacts:\n      - name: src\n        path: /mnt/src\n        git:\n          repo: \"{{workflow.parameters.source}}\"\n    outputs:\n      artifacts:\n      - name: model\n        path: /mnt/output\n        optional: true\n        archive:\n          none: {}\n    container:\n      image: tensorflow/tensorflow:latest\n      command: [sh,-c]\n      args: [\"{{workflow.parameters.command}}\"]\n      workingDir: /mnt/src\n      volumeMounts:\n      - name: data\n        mountPath: /mnt/data\n      - name: output\n        mountPath: /mnt/output\n  - name: slack-notify-success\n    container:\n      image: technosophos/slack-notify\n      command: [sh,-c]\n      args: ['SLACK_USERNAME=Worker SLACK_TITLE=\"{{workflow.name}} {{inputs.parameters.status}}\" SLACK_ICON=https://www.gravatar.com/avatar/5c4478592fe00878f62f0027be59c1bd SLACK_MESSAGE=$(cat /tmp/metrics.json)} ./slack-notify']\n    inputs:\n      parameters:\n      - name: status\n      artifacts:\n      - name: metrics\n        path: /tmp/metrics.json\n        optional: true\n"
}
```

**Comment by rushtehrani** [comment]
> ```json
>  "value": "name: command\norder: 1\nvalue: python mnist/main.py --epochs=5\n",
> ```
@aleksandrmelnikov this value seems off, is it correct?

**Comment by aleksandrmelnikov** [comment]
> > ```json
> >  "value": "name: command\norder: 1\nvalue: python mnist/main.py --epochs=5\n",
> > ```
> 
> @aleksandrmelnikov this value seems off, is it correct?

It's the raw return from the manifest parsing.
I can do more work to extract the "value:" piece itself.
I left it like that for feedback.

**Comment by rushtehrani** [comment]
It seems to be a repeat of the param itself, which suggests the value is getting parsed incorrectly, here is the actually value key/pair in the manifest:

`value: \"python mnist/main.py --epochs=5\"`
